site2/docs/cookbooks-message-throttling.md

```diff
@@ -0,0 +1,86 @@
+
+> If set ` dispatchThrottlingRateInMsg` and ` dispatchThrottlingRateInByte` both, then the message delivery is required 
+> at the same time satisfy the rules of the two, in other words the two rules are not mutually exclusive.
+
+### How it works
+Message dispatch throttling works divided into these steps:
+1. Calculate the number of messages or bytes to be delivered.
+2. Estimate the amount of data to be read from the Storage (This estimate logic is not accurate, as described below).
+3. Update the counter of message dispatch throttling.
+4. Actually deliver the message to the client.
+
+> If the quota in the current rate limiting period is not used up, the quota will not be used in the next rate limiting
+> period. However, if the quota in the current rate limiting period is exceeded, the quota in the next rate limiting
+> period is reduced. For example, if the rate-limiting rule is set to `10 /s`, `11` messages are delivered to the client
+> in the first rate-limiting period,`9` messages will be delivered to the client in the next rate-limiting period.
+
+#### Estimate the amount of data to be read from the Storage
+Messages is stored in blocks. If batch-send is not enabled, each message is packaged into an independent data block.And
+if batch-send is enabled, a batch of messages are packaged into one data block, then we can not determine the true count
+of messages in one data block; We also cannot determine the true size of this data block until we have actually read the
+it from Storage.So in order to satisfy the constraints `dispatchThrottlingRateInMsg` and `dispatchThrottlingRateInByte`,
+requires a mechanism to estimate how much need to read a data block.
+
+> If enabled the feature ` dispatchThrottlingOnBatchMessageEnabled`, instead of counting by message, we're counting
+> by blocks of data.
+
+When delivering a message to the client, the Broker looks at the size and number of messages per data block, calculates
+an average and caches it in memory, using this average to estimate how many data blocks to be read. What if there is no
+estimated value in memory when the first reading of a new topic? The current workaround is that just delivers one data
+block to clients if it is the first delivery of a new topic. Of course, the broker does not store all the information
+of data blocks for a long time to process the average accurately, which would waste memory. Instead, a compromise
+algorithm 'avg = (history-avg * 0.9 + new-avg * 0.1)' is used.
+
+### Only three granularity supports
+- Throttling limit per broker: Often applied to overload protection of broker or storage. If only to prevent the
+  excessive read requests results in Storage can not work well, you can disabled the feature
+  `dispatchThrottlingOnNonBacklogConsumerEnabled` (default is enabled). After disabled this feature, if all consumers
+  in one subscription have no backlog(it is clear that almost all read requests can hit the cache), then message
```

Comment:
How do `dispatchThrottlingRateInMsg` and `dispatchThrottlingRateInByte` work together here?
Even if we set `dispatchThrottlingRateInMsg` and `dispatchThrottlingRateInByte`, throttling won't work because there is no backlog?

Could we clarify?

Reply:
> How do dispatchThrottlingRateInMsg and dispatchThrottlingRateInByte work together here? Even if we set dispatchThrottlingRateInMsg and dispatchThrottlingRateInByte, throttling won't work because there is no backlog?

yes, it will not works